packages/docs/docs/tutorials/sidekickTutorial/part5-genUI.md

```diff
@@ -2,18 +2,149 @@
+- `<Citation>` - A citation that includes a title and a link.
+- `<NextStepsButton>` - A button that, when pressed, sends the contents of the button text
+  as the next query to the Sidekick.
+
+As long as the `outputFormat` parameter to the `<Sidekick>` component is set to `text/mdx` ,
+there is nothing more you need to do to get the Sidekick to use these standard components.
+
+## Helping the Sidekick know when to use GenUI components
+
+Even though the `<Card>` , `<Citation>` , and `<NextStepsButton>` components are available
+to the Sidekick, it won't always use them. It is often helpful to give the Sidekick some
+additional examples in terms of when it should use these UI elements.
+
+For this, we are going to add a new `<SystemMessage>` that we'll pass to the Sidekick, giving
+it some concrete examples of when to use them.
+
+```tsx
+export const finalSystemMessageBeforeResponse = (
+  <SystemMessage>
+    Respond with a `Card`. If your API call produced a 4xx error, see if you can fix the request and try again.
+    Otherwise: Give the user suggested next queries, using `NextStepsButton`. Only suggest things you can actually do.
+    Here's an example of what the final outcome should look like:
+    {`
+        <NextStepsButton prompt='See more about this function' />
+        <NextStepsButton prompt='See all of the stars for this repository' />
+        `}
+    When you give next steps, phrase them as things the user would say to you. Also, only give next steps that are fully
+    actionable by you. You cannot call any write APIs, so do not make suggestions like `create a repository`.
+  </SystemMessage>
+);
+```
+
+To use this in our Sidekick, we're going to pass it in as a `finalSystemMessageBeforeResponse` ,
+like so:
+
+```tsx
+<Sidekick
+  role="GitHub assistant"
+  systemMessage={systemMessage}
+  tools={tools}
+  finalSystemMessageBeforeResponse={finalSystemMessageBeforeResponse}
+/>
+```
+
+## Adding custom GenUI components
+
+It is possible to get the Sidekick to use your own, custom MDX components in its output.
+The challenge here is ensuring that the rendering front-end to your Sidekick incorporates
+those additional components. The generic Sidekick UI provided by Fixie is not yet
+extensible, so you would need to build your own UI frontend that renders the MDX generated
+by the Sidekick, using the GenUI components you provide. This is left as an exercise
+for the reader!
+
+To add your own GenUI components to the Sidekick, you can provide
+the optional parameters `genUIComponentNames` and `genUIExamples` to the `<Sidekick>`
+
+component, like so:
+
+```tsx
+const genUIComponentNames = [ 'MyTable' ];
+const genUIExamples = (
+  return <>
+    Whenever responding with tabular data, the {'<MyTable />'} component. Its interface is:
+    {`
+    interface MyTable {
+      header: string
+      columnNames: string[]
+      rowData: string[][]
+    }
+    `}
+  </>
```

Comment:
Minor: it would be nice to have a component that's not so close to something that already exists in Markdown.